### DISCUSSION TITLE
`Legacy`/`SegwitV0`/`Taproot` identifiers - can we do better?

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1762] [Q&A]
We are in the process of separating out types that relate to the three transaction types described by "legacy", "segwit v0", "taproot" (segwit v1). Are there better identifiers

- segwit v0
  - it is annoyingly long to write
  - it is not uniform with segwit v1 (we use "tap" or "taproot")
  - potential for confusion between wrapped and native?
- Taproot
  - Has two forms "tap" and "taproot"
  - Awesome because its short and easy to type, especially "tap"
  - non-uniform with segwit v0
  - what will next segwit version be, will it be taproot still but different?

So my question is: Can we come up with something better than `SegwitV0`?

Ideas
- `Native` (potentially "nat" and "native")

**Comment by tcharding** [comment]
Is "leg" and "legacy" going too far? I.e., we have `TapSighashType`, should the legacy one be `LegSighashType`?

**Reply by apoelstra** [reply]
I think "leg" is going too far.

Unforutnately I think `Legacy` and `SegwitV0` are the best names for the old-school types of addresses.

**Reply by tcharding** [reply]
> I think "leg" is going too far.

Came to the same conclusion while working on #1606.

> Unforutnately I think Legacy and SegwitV0 are the best names for the old-school types of addresses.

Will change #1606 back to use `SegwitV0` - tomorrow though :)

**Comment by Kixunil** [comment]
I wouldn't mind calling everything `Taproot` rather than `Tap`. Reading > writing.

Do you have specific examples where wrapped and native segwit would differ in a way we should represent it in a type?

**Reply by tcharding** [reply]
> I wouldn't mind calling everything Taproot rather than Tap. Reading > writing.

Cool, after the comments above on "leg"/"legacy" I came to this conclusion myself. Will do later after we split the ECDSA types up.

**Reply by tcharding** [reply]
> Do you have specific examples where wrapped and native segwit would differ in a way we should represent it in a type?

At this stage I do not, I just wanted to put all the pros and cons I could think of.

**Reply by Kixunil** [reply]
Now that I think about it I'd probably make exception for `TapBranchTag` (because it's referred to as such in the BIP) and definitely `tapscript`.

**Reply by tcharding** [reply]
I'm going to hold off renaming taproot types until I've spent more time on that part of the code (something I've been meaning to do for a while now).

**Comment by tcharding** [comment]
Related:

- https://github.com/murchandamus/bips/pull/1
- https://github.com/rust-bitcoin/rust-bitcoin/issues/1824
